### PR TITLE
Download Nginx Tarball from OpenWhisk CLI release, rather than constructing file-by-file

### PR DESCRIPTION
**Pull request by jonpspri**
Ansible build of Nginx edge now downloads the Openwhisk CLI tarball from the GitHub release repository rather than trying to download for individual architectures and then piece together the resulting index.  By taking this approach, we've eliminated a circular dependency where the `incubator-openwhisk` project had to be aware of all the os/architecture combinations supported by the CLI.

### CODE REVIEW

File: ansible/roles/invoker/tasks/deploy.yml [hunk 1]
```diff
@@ -129,6 +129,12 @@
   set_fact:
     invoker_args: "{{ invoker.arguments }} {{ invoker.jmxremote.jvmArgs }}"
 
+- name: stop and remove the invoke container if it's currently running
+  docker_container:
```
Comment (rabbah) [inline ansible/roles/invoker/tasks/deploy.yml]: these changes are orthogonal to the pr no?

Comment (jonpspri) [reply ansible/roles/invoker/tasks/deploy.yml]: Oops, yes.  They snuck in because I got tired of manually stopping and removing the container when running multiple tests.  I'll pull them from this branch and file a separate PR.

File: ansible/roles/cli/tasks/deploy.yml [hunk 2]
```diff
@@ -1,29 +1,81 @@
...
+#  default) located in the /tmp/... directory tree, which has a sticky bit
+#  set.  Said sticky bit creates no end of troubles for tar, so we're going
+#  to just avoid it entirely, rather than muck about with who's got which tar
+#  with which right tar options installed where.  It makes for many more
+#  items in this ansible playbook than we'd hoped, but at least it's (fairly)
+#  straightforward.
+#
+- name: "Download release archive to build directory ..."
+  get_url:
+    url: "{{ openwhisk_cli.remote.location }}/{{ openwhisk_cli.archive_name}}-{{ openwhisk_cli_tag }}-all.tgz"
```
Comment (dubeejw) [inline ansible/roles/cli/tasks/deploy.yml]: @mdeuser, note this change. Instead of downloading individual compressed files for each platform, we will now download one zip that contains all the needed binaries.

Comment (houshengbo) [reply ansible/roles/cli/tasks/deploy.yml]: It is a good idea to download all at once to reduce network connections.
However, there is one remark I want to make: we should give the users the options to choose which type of architectures and operating system. By default, we can set to ALL, and download all the binaries in side one zip. Users should also be able to select the specified binaries, based on the configurations.
In summary, I support to add this PR in terms of downloading all the binaries in one zip, but also keep the old mechanism to download them one by one. Add some configuration option to decide ALL or some binaries.

**Comment by rabbah** [comment]
How does this affect bundling a locally built CLI?

File: ansible/roles/cli/tasks/deploy.yml (hunk 2 above)
Comment (jonpspri) [reply ansible/roles/cli/tasks/deploy.yml]: Why?  We'd be adding significant complexity to what is already a complex local build:

1.  Adjustment/re-generation of content.json
2.  Specification and parsing of local build architecture(s)
3.  Potentially not having the binary that's needed for the remainder of setup

For what benefit?  Saving a few megabytes of disk space and trimming a directory structure that is unseen.  Users running the CLI still have the option of downloading the appropriate architecture from the nginx server that's running.  So I suppose my core question is "who are 'users' in the conversation above"?

I'll do the change if it's what's needed to get this off my plate, but the whole rationale was to avoid complicating the build and to avoid having a circular dependency where the list of architectures was to be kept in the `incubator-openwhisk` repository.  If we're doing individual downloads, I'd abandon the archive download regardless of network connections.

**Comment by jonpspri** [comment]
@rabbah For a locally build API, the archive is sourced from the local build.  The flags to declare which approach to take remain.

File: ansible/roles/cli/tasks/deploy.yml [hunk 3]
```diff
@@ -1,30 +1,88 @@
...
+    dest: "{{ openwhisk_build_dir }}/{{ openwhisk_cli.archive_name }}.tgz"
+    headers: "{{ openwhisk_cli.headers | default('') }}"
```
Comment (dubeejw) [inline ansible/roles/cli/tasks/deploy.yml]: This needs to be `openwhisk_cli.remote.headers`.

**Comment by houshengbo** [comment]
Still failed the test case:
services.KafkaConnectorTests > Kafka connector should send and receive a kafka message even after session timeout FAILED
    org.scalatest.exceptions.TestFailedException: 1 was not equal to 2
        at org.scalatest.MatchersHelper$.indicateFailure(MatchersHelper.scala:340)
        at org.scalatest.Matchers$AnyShouldWrapper.shouldBe(Matchers.scala:6864)
        at services.KafkaConnectorTests$$anonfun$2$$anonfun$apply$mcV$sp$3.apply(KafkaConnectorTests.scala:128)
        at services.KafkaConnectorTests$$anonfun$2$$anonfun$apply$mcV$sp$3.apply(KafkaConnectorTests.scala:125)
        at scala.collection.immutable.Range.foreach(Range.scala:160)
        at services.KafkaConnectorTests$$anonfun$2.apply$mcV$sp(KafkaConnectorTests.scala:125)
        at services.KafkaConnectorTests$$anonfun$2.apply(KafkaConnectorTests.scala:120)
        at services.KafkaConnectorTests$$anonfun$2.apply(KafkaConnectorTests.scala:120)
        at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
        at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
        at org.scalatest.Transformer.apply(Transformer.scala:22)
        at org.scalatest.Transformer.apply(Transformer.scala:20)
        at org.scalatest.FlatSpecLike$$anon$1.apply(FlatSpecLike.scala:1682)
        at org.scalatest.TestSuite$class.withFixture(TestSuite.scala:196)
        at org.scalatest.FlatSpec.withFixture(FlatSpec.scala:1685)
        at org.scalatest.FlatSpecLike$class.invokeWithFixture$1(FlatSpecLike.scala:1679)
        at org.scalatest.FlatSpecLike$$anonfun$runTest$1.apply(FlatSpecLike.scala:1692)
        at org.scalatest.FlatSpecLike$$anonfun$runTest$1.apply(FlatSpecLike.scala:1692)
        at org.scalatest.SuperEngine.runTestImpl(Engine.scala:289)
        at org.scalatest.FlatSpecLike$class.runTest(FlatSpecLike.scala:1692)
        at org.scalatest.FlatSpec.runTest(FlatSpec.scala:1685)
        at org.scalatest.FlatSpecLike$$anonfun$runTests$1.apply(FlatSpecLike.scala:1750)
        at org.scalatest.FlatSpecLike$$anonfun$runTests$1.apply(FlatSpecLike.scala:1750)
        at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:396)
        at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:384)
        at scala.collection.immutable.List.foreach(List.scala:392)
        at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
        at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:373)
        at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:410)
        at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:384)
        at scala.collection.immutable.List.foreach(List.scala:392)
        at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
        at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:379)
        at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:461)
        at org.scalatest.FlatSpecLike$class.runTests(FlatSpecLike.scala:1750)
        at org.scalatest.FlatSpec.runTests(FlatSpec.scala:1685)
        at org.scalatest.Suite$class.run(Suite.scala:1147)
        at org.scalatest.FlatSpec.org$scalatest$FlatSpecLike$$super$run(FlatSpec.scala:1685)
        at org.scalatest.FlatSpecLike$$anonfun$run$1.apply(FlatSpecLike.scala:1795)
        at org.scalatest.FlatSpecLike$$anonfun$run$1.apply(FlatSpecLike.scala:1795)
        at org.scalatest.SuperEngine.runImpl(Engine.scala:521)
        at org.scalatest.FlatSpecLike$class.run(FlatSpecLike.scala:1795)
        at services.KafkaConnectorTests.org$scalatest$BeforeAndAfterAll$$super$run(KafkaConnectorTests.scala:41)
        at org.scalatest.BeforeAndAfterAll$class.liftedTree1$1(BeforeAndAfterAll.scala:213)
        at org.scalatest.BeforeAndAfterAll$class.run(BeforeAndAfterAll.scala:210)
        at services.KafkaConnectorTests.run(KafkaConnectorTests.scala:41)

**Comment by jonpspri** [comment]
Closing and reopening to re-execute Travis.  I suspect the Kafka test case is a Heisenbug.

**Comment by dubeejw** [comment]
PG3 1937 🔵

File: ansible/edge.yml [hunk 4]
```diff
@@ -1,8 +1,8 @@
 ---
 # This playbook deploys Openwhisk Edge servers.
 # The edge is usually populated with NGINX serving as proxy.
-# The CLI also gets built and published for downloading from NGINX.  
-# SDKs for blackbox and iOS get published to NGINX also.
+# The CLI also gets built and published for downloading from NGINX.
+# SDKs for blackbox and iOS get published to NGINX also ( )
```
Comment (dubeejw) [inline ansible/edge.yml]: What is `( )`?

Comment (jonpspri) [reply ansible/edge.yml]: Text editor artifact.  Maybe a leftover from my not being sure where the SDKs come from.
